Mathlib/Algebra/CharP/ExpChar.lean

```diff
@@ -111,3 +112,79 @@ end NoZeroDivisors
 end Nontrivial
 
 end Semiring
+
+theorem ExpChar.exists [Ring R] [NoZeroDivisors R] [Nontrivial R] :
+    ∃ q, ExpChar R q := by
+  obtain ⟨p, h⟩ := CharP.exists R
+  by_cases hp : p = 0
+  · exact ⟨1, by rw [hp] at h; haveI := CharP.charP_to_charZero R; exact .zero⟩
+  exact ⟨p, haveI := NeZero.mk hp; .prime (CharP.char_is_prime_of_pos R p).out⟩
+
+/-- If the algebra map `R →+* A` is injective then `A` has the same exponential characteristic
+as `R`. -/
+theorem expChar_of_injective_algebraMap {R A : Type*}
+    [CommSemiring R] [Semiring A] [Algebra R A] (h : Function.Injective (algebraMap R A))
+    (q : ℕ) [hR : ExpChar R q] : ExpChar A q := by
+  cases' hR with _ _ hprime _
+  · haveI := charZero_of_injective_algebraMap h; exact .zero
+  haveI := charP_of_injective_algebraMap h q; exact .prime hprime
+
+theorem add_pow_expChar_of_commute [Semiring R] {q : ℕ} [hR : ExpChar R q]
+    (x y : R) (h : Commute x y) : (x + y) ^ q = x ^ q + y ^ q := by
+  cases' hR with _ _ hprime _
+  · simp only [pow_one]
+  haveI := Fact.mk hprime; exact add_pow_char_of_commute R x y h
+
+theorem add_pow_expChar_pow_of_commute [Semiring R] {q : ℕ} [hR : ExpChar R q]
+    {n : ℕ} (x y : R) (h : Commute x y) : (x + y) ^ (q ^ n) = x ^ (q ^ n) + y ^ (q ^ n) := by
+  cases' hR with _ _ hprime _
+  · simp only [one_pow, pow_one]
+  haveI := Fact.mk hprime; exact add_pow_char_pow_of_commute R x y h
+
+theorem sub_pow_expChar_of_commute [Ring R] {q : ℕ} [hR : ExpChar R q]
+    (x y : R) (h : Commute x y) : (x - y) ^ q = x ^ q - y ^ q := by
+  cases' hR with _ _ hprime _
+  · simp only [pow_one]
+  haveI := Fact.mk hprime; exact sub_pow_char_of_commute R x y h
+
+theorem sub_pow_expChar_pow_of_commute [Ring R] {q : ℕ} [hR : ExpChar R q]
+    {n : ℕ} (x y : R) (h : Commute x y) : (x - y) ^ (q ^ n) = x ^ (q ^ n) - y ^ (q ^ n) := by
+  cases' hR with _ _ hprime _
+  · simp only [one_pow, pow_one]
+  haveI := Fact.mk hprime; exact sub_pow_char_pow_of_commute R x y h
+
+theorem add_pow_expChar [CommSemiring R] {q : ℕ} [hR : ExpChar R q]
+    (x y : R) : (x + y) ^ q = x ^ q + y ^ q := by
+  cases' hR with _ _ hprime _
+  · simp only [pow_one]
+  haveI := Fact.mk hprime; exact add_pow_char R x y
```

Comment:
Neg, Inv, Div (for DivisionRing/Field) lemmas are missing but they are also missing for CharP, so maybe don't add them for now. We should also consider redefining [frobenius](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#frobenius) using ExpChar; I think getting rid of `Fact p.prime` is definitely nice.

Reply:
> Neg, Inv, Div (for DivisionRing/Field) lemmas are missing but they are also missing for CharP, so maybe don't add them for now.

For `Neg` it is a corollary of `neg_one_pow_expChar` and `neg_one_pow_expChar_pow`. For `Inv` do you mean `(x^{-1})^q=(x^q)^{-1}`? I think it's true independent of the characteristic, so probably it's in an earlier file.

Reply:
Indeed there's already [pow_inv_comm₀](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Power.html#pow_inv_comm%E2%82%80). Sub is (can be) a corollary of Add and Neg, Neg is a corollary of Mul and Neg_one, and Div is a corollary of Mul and Inv, but they all can be useful. But they are not strictly necessary as we can just [`rw [frobenius_def]`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#frobenius_def) and use the RingHom lemmas ([map_div₀](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupWithZero/Units/Lemmas.html#map_div%E2%82%80) etc. already exists).

Reply:
We also have [iterate_frobenius](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#iterate_frobenius) but the problem problem is that Lean won't recognize `(frobenius R p)^[n]` as a RingHom, so we'd need to introduce a `iterate_frobenius` (or `frobenius_pow`?) RingHom if we want to get rid of these lemmas.

Reply:
The point is that in the application it is written directly as `x ^ (q ^ n)` if `q` is the exponential character, for example:
```lean
/-- A field extension `E / F` of exponential characteristic `q` is purely inseparable
if and only if for every element `x` of `E`, there exists a natural number `n` such that
`x ^ (q ^ n)` is contained in `F`. -/
theorem isPurelyInseparable_iff_mem_pow (q : ℕ) [hF : ExpChar F q] :
    IsPurelyInseparable F E ↔ ∀ x : E, ∃ n : ℕ, x ^ (q ^ n) ∈ (algebraMap F E).range := by ...
```
It's written directly as `x ^ (q ^ n)` without using `frobenius` :-/